Getting Started with the C/C++ Chainguard Image (#1723)
chainguard-dev/edu: content/chainguard/chainguard-images/getting-started/c/index.md

Let us begin by creating a Dockerfile called `Dockerfile1` for our image. Execute the following code with the text editor of your choice to begin editing a new file.

```sh
nano Dockerfile1
```

The following Dockerfile will:
1. Use the `gcc-glibc:latest` Chainguard Image as the base image;
2. Set the current working directory to `/usr/bin`;
3. Copy our `hello.c` program code to the current directory;
4. Compile our program and name it `hello`;
5. Execute the compiled binary when the container is started.

```Dockerfile
# Example 1 - Single Stage Build for C

FROM cgr.dev/chainguard/gcc-glibc:latest

WORKDIR /usr/bin

COPY hello.c ./

RUN ["gcc", "-Wall", "-o", "hello", "hello.c"]

ENTRYPOINT ["./hello"]
```

Copy the text in to your Dockerfile, save, and close it.

> Next, start Docker Engine on your machine. Execute the following command in your terminal. The `-f` flag specifies the Dockerfile which we are using to build from, and the `-t` flag will tag our image with a meaningful name.

Author: I might be missing something... when I run any docker-related commands, they fail unless I manually start up Docker by opening Docker Desktop. I'll have to work out why that is 😓

### Step 1: Setting up a Demo Application

In your terminal, create a new file called `hello.cpp` with the text editor of your choice.

```sh
nano hello.cpp
```

Copy the following C++ code to the file you just created. This code will display a simple greeting specifying that it was written in C++.

```C++
/* Chainguard Academy (edu.chainguard.dev)
*  Getting Started with the C/C++ Chainguard Images
*  Example 3 - C++
*/

#include <iostream>
using namespace std;

// Main Function
int main(){
    cout << "Hello, world!\n";
    cout << "I am a demo from the Chainguard Academy.\n";
    cout << "My code was written in C++.\n";

    return 0;
}
```

When you are done editing your file, save and close it.

You can now compile your C++ program using `g++`. Execute the following command in your terminal to compile the program. The command will display any compiler warnings or errors and will name the resultant binary `hello`.

```sh
g++ -Wall -o hello hello.cpp
```

Now you can test your compiled binary.

```sh
./hello
```

You should see the following output in your terminal.

```Output
Hello, world!
I am a demo from the Chainguard Academy.
My code was written in C++.
```

> 

Reviewer: again, could you add a transition/closing sentence here?

### Step 2: Creating the Dockerfile

> Now, with a working C++ example, we can compile and run our program using a multi-stage build. With the text editor of your choice, create a new file named `Dockerfile3`.

Reviewer: ```suggestion
With a working C++ example, you can compile and run our program using a multi-stage build. With the text editor of your choice, create a new file named `Dockerfile3`.
```

```sh
nano Dockerfile3
```

The following Dockerfile will:

> 1. Use the `gcc-glibc` Chainguard Image as the builder stage;

Reviewer: I wanted to say: nice attention to detail with these lists! They're all formatted beautifully and identically :ok_hand:

2. Set the current working directory of the image to `/usr/bin`;
3. Copy our `hello.cpp` program code to the current directory;
4. Compile the program using `g++` and name it `hello`;
5. Begin a new stage using the `glibc-dynamic` Chainguard Image;
6. Set the current working directory of the new image to `/usr/bin`;
7. Copy the compiled `hello` C++ binary from our builder stage;
8. Execute our binary from the `glibc-dynamic` image when the container is started.

```Dockerfile3
# Example 3 - Multi-Stage Build for C++

FROM cgr.dev/chainguard/gcc-glibc:latest AS builder

WORKDIR /usr/bin

COPY hello.cpp ./

RUN ["g++", "-Wall", "-o", "hello", "hello.cpp"]

FROM cgr.dev/chainguard/glibc-dynamic:latest

WORKDIR /usr/bin

COPY --from=builder /usr/bin/hello ./

ENTRYPOINT ["./hello"]
```

When you are finished editing your Dockerfile, save and close it.

> With our new Dockerfile created, we can build the image. Execute the following command in your terminal to build your multi-stage C++ image.

Reviewer: ```suggestion
With your new Dockerfile created, you can build the image. Execute the following command in your terminal to build your multi-stage C++ image.
```

```sh
docker build -f Dockerfile3 -t example3:latest .
```

> With your image built, we can now run it with the following command.

Reviewer: ```suggestion
With your image built, you can now run it with the following command.
```

I feel like I might have missed some paragraphs that split between "you" and "we" - maybe do a scan for this?

```
docker run --name example3 example3:latest
```

> You should see output in your terminal identical to that of the C++ binary you compiled locally.

Reviewer: ```suggestion
You will see output in your terminal identical to that of the C++ binary you compiled locally.
```

```Output
Hello, world!
I am a demo from the Chainguard Academy.
My code was written in C++.

> ```

Reviewer: again, add a transition sentence after this output
